Source/AGXUnreal/Public/Terrain/AGX_MovableTerrainComponent.h

```diff
@@ -0,0 +1,300 @@
+#pragma once
+
+// AGX Dynamics for Unreal include.s
+#include "AGX_NativeOwner.h"
+#include "Terrain/TerrainBarrier.h"
+#include "Terrain/AGX_TerrainMeshUtilities.h"
+#include "AGX_ShovelReference.h"
+
+#include "CoreMinimal.h"
+#include "ProceduralMeshComponent.h"
+#include "AGX_MovableTerrainComponent.generated.h"
+
+
+class UAGX_ShapeComponent;
+class UAGX_TerrainMaterial;
+class UAGX_ShovelComponent;
+class UNiagaraSystem;
+class UNiagaraComponent;
+
+struct MeshTile
+{
+	FVector2D Center;
+	FVector2D Size;
+	FIntVector2 Resolution;
+	MeshTile(FVector2D TileCenter, FVector2D TileSize, FIntVector2 TileRes)
+	{
+		Center = TileCenter;
+		Size = TileSize;
+		Resolution = TileRes;
+	}
+};
+
+/**
+ *
+ */
+UCLASS(
+	ClassGroup = "AGX_Terrain", meta = (BlueprintSpawnableComponent),
+	Hidecategories = (Cooking, Collision, Input, LOD, Physics, Replication))
+class AGXUNREAL_API UAGX_MovableTerrainComponent : public UProceduralMeshComponent
+{
+	GENERATED_BODY()
+public:
+	UAGX_MovableTerrainComponent(const FObjectInitializer& ObjectInitializer);
+
+	bool HasNative() const;
+	void CreateNative();
+	FTerrainBarrier* GetNative();
+	const FTerrainBarrier* GetNative() const;
+
+	virtual void BeginPlay() override;
+	virtual void EndPlay(const EEndPlayReason::Type EndPlayReason) override;
+
+	virtual void TickComponent(
+		float DeltaTime, ELevelTick TickType, FActorComponentTickFunction* ThisTickFunction);
+
+#if WITH_EDITOR
+	virtual void PostInitProperties() override;
+	virtual void PostEditChangeChainProperty(FPropertyChangedChainEvent& Event) override;
+#endif
+
+protected:
+	UPROPERTY(EditAnywhere, Category = "AGX Terrain Editor")
+	bool bRebuildMesh = false;
+
+	UPROPERTY(
+		EditAnywhere, Category = "AGX Terrain Shape", 
+		BlueprintReadWrite, Meta = (ExposeOnSpawn))
+	FVector2D Size = FVector2D(200.0f, 200.0f);
+
+	UPROPERTY(
+		EditAnywhere, Category = "AGX Terrain Shape",
+		BlueprintReadWrite, 
+		Meta = (ExposeOnSpawn, ClampMin = "1", UIMin = "1", ClampMax = "100", UIMax = "100"))
+	double ElementSize = 10;
```

Comment:
For properties that should not be editable in runtime, you can override the `CanEditChange` and handle that there. That way they will be grayed out in the Details Panel during Play.

You can sneek peak on other Components for inspiration or ping me if it is unclear how.

There's a few things to consider regarding this, whether or not to expose functions to Blueprints etc, perhaps we should have a quick walkthrough about all of this in a meet? So that it doesn't become overwhelming.